lib/matplotlib/figure.py

```diff
@@ -52,6 +52,71 @@ def _stale_figure_callback(self, val):
         self.figure.stale = val
 
 
+class _AxesStack(cbook.Stack):
+    """
+    Specialization of Stack, to handle all tracking of Axes in a Figure.
+
+    This stack stores ``ind, axes`` pairs, where ``ind`` is a serial index
+    tracking the order in which axes were added.
+
+    AxesStack is a callable; calling it returns the current axes.
+    """
+
+    def __init__(self):
+        super().__init__()
+        self._ind = 0
+
+    def as_list(self):
+        """
+        Return a list of the Axes instances that have been added to the figure.
+        """
+        return [a for i, a in sorted(self._elements)]
+
+    def _entry_from_axes(self, e):
+        return next(((ind, a) for ind, a in self._elements if a == e), None)
+
+    def remove(self, a):
+        """Remove the axes from the stack."""
+        super().remove(self._entry_from_axes(a))
+
+    def bubble(self, a):
+        """
+        Move the given axes, which must already exist in the stack, to the top.
+        """
+        return super().bubble(self._entry_from_axes(a))
+
+    def add(self, a):
+        """
+        Add Axes *a* to the stack, and return the stack.
+
+        If *a* is already on the stack, don't add it again, but
+        return *None*.
+        """
+        # All the error checking may be unnecessary; but this method
+        # is called so seldom that the overhead is negligible.
+        _api.check_isinstance(Axes, a=a)
+
+        if a in self:
+            return None
```

Comment:
probably should test this since it seems someone could change this?

Reply:
I'll check further, but this may not ever happen, as the `Figure.add_*` might handle that already now.

Reply:
There are 3 effects of this early `return`:
1. A different return value, but `_AxesStack.add` is only called from `_add_axes_internal`, and that doesn't care about the return value.
2. The `Axes` doesn't get added to the stack multiple times, but that's what we want anyway.
3. The `Axes` would not be made current (i.e., moved to the end), but there's a `sca` in `_add_axes_internal`, so we don't need to worry about that.

I thought some tests checked `add_axes`/`add_subplot` with an existing `Axes`, but it appears not if this isn't getting coverage. I will add some more tests and remove the return value, which no-one uses.